File: output/csharp/tests/moore_machine.cs
using System;
using System.Collections.Generic;


using System;
using System.Collections.Generic;

// Moore Machine - output depends ONLY on state (output on entry)
// See: https://www.geeksforgeeks.org/mealy-and-moore-machines-in-toc/

class MooreMachineFrameEvent {
    public string _message;
    public Dictionary<string, object> _parameters;

    public MooreMachineFrameEvent(string message) {
        this._message = message;
        this._parameters = null;
    }

    public MooreMachineFrameEvent(string message, Dictionary<string, object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class MooreMachineFrameContext {
    public MooreMachineFrameEvent _event;
    public object _return;
    public Dictionary<string, object> _data;

    public MooreMachineFrameContext(MooreMachineFrameEvent ev, object defaultReturn) {
        this._event = ev;
        this._return = defaultReturn;
        this._data = new Dictionary<string, object>();
    }
}

class MooreMachineCompartment {
    public string state;
    public Dictionary<string, object> state_args;
    public Dictionary<string, object> state_vars;
    public Dictionary<string, object> enter_args;
    public Dictionary<string, object> exit_args;
    public MooreMachineFrameEvent forward_event;
    public MooreMachineCompartment parent_compartment;

    public MooreMachineCompartment(string state) {
        this.state = state;
        this.state_args = new Dictionary<string, object>();
        this.state_vars = new Dictionary<string, object>();
        this.enter_args = new Dictionary<string, object>();
        this.exit_args = new Dictionary<string, object>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    public MooreMachineCompartment Copy() {
        MooreMachineCompartment c = new MooreMachineCompartment(this.state);
        c.state_args = new Dictionary<string, object>(this.state_args);
        c.state_vars = new Dictionary<string, object>(this.state_vars);
        c.enter_args = new Dictionary<string, object>(this.enter_args);
        c.exit_args = new Dictionary<string, object>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class MooreMachine {
    private List<MooreMachineCompartment> _state_stack;
    private MooreMachineCompartment __compartment;
    private MooreMachineCompartment __next_compartment;
    private List<MooreMachineFrameContext> _context_stack;
    public int current_output = 0;

    public MooreMachine() {
        _state_stack = new List<MooreMachineCompartment>();
        _context_stack = new List<MooreMachineFrameContext>();
        __compartment = new MooreMachineCompartment("Q0");
        __next_compartment = null;
        MooreMachineFrameEvent __frame_event = new MooreMachineFrameEvent("$>");
        MooreMachineFrameContext __ctx = new MooreMachineFrameContext(__frame_event, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    private void __kernel(MooreMachineFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            MooreMachineFrameEvent exit_event = new MooreMachineFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                MooreMachineFrameEvent enter_event = new MooreMachineFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message == "$>") {
                    __router(forward_event);
                } else {
                    MooreMachineFrameEvent enter_event = new MooreMachineFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(MooreMachineFrameEvent __e) {
        string state_name = __compartment.state;
        if (state_name == "Q0") {
            _state_Q0(__e);
        } else if (state_name == "Q1") {
            _state_Q1(__e);
        } else if (state_name == "Q2") {
            _state_Q2(__e);
        } else if (state_name == "Q3") {
            _state_Q3(__e);
        } else if (state_name == "Q4") {
            _state_Q4(__e);
        }
    }

    private void __transition(MooreMachineCompartment next) {
        __next_compartment = next;
    }

    public void i_0() {
        MooreMachineFrameEvent __e = new MooreMachineFrameEvent("i_0");
        MooreMachineFrameContext __ctx = new MooreMachineFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    public void i_1() {
        MooreMachineFrameEvent __e = new MooreMachineFrameEvent("i_1");
        MooreMachineFrameContext __ctx = new MooreMachineFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    private void _state_Q0(MooreMachineFrameEvent __e) {
        if (__e._message == "$>") {
            this.set_output(0);
        } else if (__e._message == "i_0") {
            { var __new_compartment = new MooreMachineCompartment("Q1");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        } else if (__e._message == "i_1") {
            { var __new_compartment = new MooreMachineCompartment("Q2");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        }
    }

    private void _state_Q1(MooreMachineFrameEvent __e) {
        if (__e._message == "$>") {
            this.set_output(0);
        } else if (__e._message == "i_0") {
            { var __new_compartment = new MooreMachineCompartment("Q1");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        } else if (__e._message == "i_1") {
            { var __new_compartment = new MooreMachineCompartment("Q3");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        }
    }

    private void _state_Q2(MooreMachineFrameEvent __e) {
        if (__e._message == "$>") {
            this.set_output(0);
        } else if (__e._message == "i_0") {
            { var __new_compartment = new MooreMachineCompartment("Q4");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        } else if (__e._message == "i_1") {
            { var __new_compartment = new MooreMachineCompartment("Q2");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        }
    }

    private void _state_Q3(MooreMachineFrameEvent __e) {
        if (__e._message == "$>") {
            this.set_output(1);
        } else if (__e._message == "i_0") {
            { var __new_compartment = new MooreMachineCompartment("Q4");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        } else if (__e._message == "i_1") {
            { var __new_compartment = new MooreMachineCompartment("Q2");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        }
    }

    private void _state_Q4(MooreMachineFrameEvent __e) {
        if (__e._message == "$>") {
            this.set_output(1);
        } else if (__e._message == "i_0") {
            { var __new_compartment = new MooreMachineCompartment("Q1");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        } else if (__e._message == "i_1") {
            { var __new_compartment = new MooreMachineCompartment("Q3");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        }
    }

    private void set_output(int value) {
                    this.current_output = value;
    }

    public int get_output() {
                    return this.current_output;
    }
}

class Program {
    static void Main(string[] args) {
        Console.WriteLine("TAP version 14");
        Console.WriteLine("1..5");

        MooreMachine m = new MooreMachine();

        // Initial state Q0 has output 0
        if (m.get_output() == 0) {
            Console.WriteLine("ok 1 - moore initial state Q0 has output 0");
        } else {
            Console.WriteLine("not ok 1 - moore initial state Q0 has output 0 # got " + m.get_output());
        }

        // i_0: Q0 -> Q1 (output 0)
        m.i_0();
        if (m.get_output() == 0) {
            Console.WriteLine("ok 2 - moore Q1 has output 0");
        } else {
            Console.WriteLine("not ok 2 - moore Q1 has output 0 # got " + m.get_output());
        }

        // i_1: Q1 -> Q3 (output 1)
        m.i_1();
        if (m.get_output() == 1) {
            Console.WriteLine("ok 3 - moore Q3 has output 1");
        } else {
            Console.WriteLine("not ok 3 - moore Q3 has output 1 # got " + m.get_output());
        }

        // i_0: Q3 -> Q4 (output 1)
        m.i_0();
        if (m.get_output() == 1) {
            Console.WriteLine("ok 4 - moore Q4 has output 1");
        } else {
            Console.WriteLine("not ok 4 - moore Q4 has output 1 # got " + m.get_output());
        }

        // i_0: Q4 -> Q1 (output 0)
        m.i_0();
        if (m.get_output() == 0) {
            Console.WriteLine("ok 5 - moore Q1 has output 0 again");
        } else {
            Console.WriteLine("not ok 5 - moore Q1 has output 0 again # got " + m.get_output());
        }

        Console.WriteLine("# PASS - Moore machine outputs depend ONLY on state");
    }
}

File: tests/common/positive/automata/moore_machine.rs

// Moore Machine - output depends ONLY on state (output on entry)
// See: https://www.geeksforgeeks.org/mealy-and-moore-machines-in-toc/

#[allow(dead_code)]
struct MooreMachineFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl Clone for MooreMachineFrameEvent {
    fn clone(&self) -> Self {
        Self {
            message: self.message.clone(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

impl MooreMachineFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
    fn new_with_params(message: &str, params: &std::collections::HashMap<String, String>) -> Self {
        Self {
            message: message.to_string(),
            parameters: params.iter().map(|(k, v)| (k.clone(), Box::new(v.clone()) as Box<dyn std::any::Any>)).collect(),
        }
    }
}

#[allow(dead_code)]
struct MooreMachineFrameContext {
    event: MooreMachineFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl MooreMachineFrameContext {
    fn new(event: MooreMachineFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
enum MooreMachineStateContext {
    Q0,
    Q1,
    Q2,
    Q3,
    Q4,
    Empty,
}

impl Default for MooreMachineStateContext {
    fn default() -> Self {
        MooreMachineStateContext::Q0
    }
}

#[allow(dead_code)]
#[derive(Clone)]
struct MooreMachineCompartment {
    state: String,
    state_context: MooreMachineStateContext,
    enter_args: std::collections::HashMap<String, String>,
    exit_args: std::collections::HashMap<String, String>,
    forward_event: Option<MooreMachineFrameEvent>,
    parent_compartment: Option<Box<MooreMachineCompartment>>,
}

impl MooreMachineCompartment {
    fn new(state: &str) -> Self {
        let state_context = match state {
            "Q0" => MooreMachineStateContext::Q0,
            "Q1" => MooreMachineStateContext::Q1,
            "Q2" => MooreMachineStateContext::Q2,
            "Q3" => MooreMachineStateContext::Q3,
            "Q4" => MooreMachineStateContext::Q4,
            _ => MooreMachineStateContext::Empty,
        };
        Self {
            state: state.to_string(),
            state_context,
            enter_args: std::collections::HashMap::new(),
            exit_args: std::collections::HashMap::new(),
            forward_event: None,
            parent_compartment: None,
        }
    }
}

#[allow(dead_code)]
pub struct MooreMachine {
    _state_stack: Vec<MooreMachineCompartment>,
    __compartment: MooreMachineCompartment,
    __next_compartment: Option<MooreMachineCompartment>,
    _context_stack: Vec<MooreMachineFrameContext>,
    pub current_output: i32,
}

#[allow(non_snake_case)]
impl MooreMachine {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            current_output: Default::default(),
            __compartment: MooreMachineCompartment::new("Q0"),
            __next_compartment: None,
        };
        let __frame_event = MooreMachineFrameEvent::new("$>");
        let __ctx = MooreMachineFrameContext::new(__frame_event, None);
        this._context_stack.push(__ctx);
        this.__kernel();
        this._context_stack.pop();
        this
    }

    fn __kernel(&mut self) {
        // Clone event from context stack (needed for borrow checker)
        let __e = self._context_stack.last().unwrap().event.clone();
        // Route event to current state
        self.__router(&__e);
        // Process any pending transition
        while self.__next_compartment.is_some() {
            let next_compartment = self.__next_compartment.take().unwrap();
            // Exit current state (with exit_args from current compartment)
            let exit_event = MooreMachineFrameEvent::new_with_params("<$", &self.__compartment.exit_args);
            self.__router(&exit_event);
            // Switch to new compartment
            self.__compartment = next_compartment;
            // Enter new state (or forward event)
            if self.__compartment.forward_event.is_none() {
                let enter_event = MooreMachineFrameEvent::new_with_params("$>", &self.__compartment.enter_args);
                self.__router(&enter_event);
            } else {
                // Forward event to new state
                let forward_event = self.__compartment.forward_event.take().unwrap();
                if forward_event.message == "$>" {
                    // Forwarding enter event - just send it
                    self.__router(&forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    let enter_event = MooreMachineFrameEvent::new_with_params("$>", &self.__compartment.enter_args);
                    self.__router(&enter_event);
                    self.__router(&forward_event);
                }
            }
        }
    }

    fn __router(&mut self, __e: &MooreMachineFrameEvent) {
        match self.__compartment.state.as_str() {
            "Q0" => self._state_Q0(__e),
            "Q1" => self._state_Q1(__e),
            "Q2" => self._state_Q2(__e),
            "Q3" => self._state_Q3(__e),
            "Q4" => self._state_Q4(__e),
            _ => {}
        }
    }

    fn __transition(&mut self, next_compartment: MooreMachineCompartment) {
        self.__next_compartment = Some(next_compartment);
    }

    fn __push_transition(&mut self, new_compartment: MooreMachineCompartment) {
        // Exit current state (old compartment still in place for routing)
        let exit_event = MooreMachineFrameEvent::new_with_params("<$", &self.__compartment.exit_args);
        self.__router(&exit_event);
        // Swap: old compartment moves to stack, new takes its place
        let old = std::mem::replace(&mut self.__compartment, new_compartment);
        self._state_stack.push(old);
        // Enter new state (or forward event) — matches kernel logic
        if self.__compartment.forward_event.is_none() {
            let enter_event = MooreMachineFrameEvent::new_with_params("$>", &self.__compartment.enter_args);
            self.__router(&enter_event);
        } else {
            let forward_event = self.__compartment.forward_event.take().unwrap();
            if forward_event.message == "$>" {
                self.__router(&forward_event);
            } else {
                let enter_event = MooreMachineFrameEvent::new_with_params("$>", &self.__compartment.enter_args);
                self.__router(&enter_event);
                self.__router(&forward_event);
            }
        }
    }

    pub fn i_0(&mut self) {
        let mut __e = MooreMachineFrameEvent::new("i_0");
        let mut __ctx = MooreMachineFrameContext::new(__e, None);
        self._context_stack.push(__ctx);
        self.__kernel();
        self._context_stack.pop();
    }

    pub fn i_1(&mut self) {
        let mut __e = MooreMachineFrameEvent::new("i_1");
        let mut __ctx = MooreMachineFrameContext::new(__e, None);
        self._context_stack.push(__ctx);
        self.__kernel();
        self._context_stack.pop();
    }

    fn _state_Q0(&mut self, __e: &MooreMachineFrameEvent) {
        match __e.message.as_str() {
            "$>" => { self._s_Q0_enter(__e); }
            "i_0" => { self._s_Q0_i_0(__e); }
            "i_1" => { self._s_Q0_i_1(__e); }
            _ => {}
        }
    }

    fn _state_Q1(&mut self, __e: &MooreMachineFrameEvent) {
        match __e.message.as_str() {
            "$>" => { self._s_Q1_enter(__e); }
            "i_0" => { self._s_Q1_i_0(__e); }
            "i_1" => { self._s_Q1_i_1(__e); }
            _ => {}
        }
    }

    fn _state_Q2(&mut self, __e: &MooreMachineFrameEvent) {
        match __e.message.as_str() {
            "$>" => { self._s_Q2_enter(__e); }
            "i_0" => { self._s_Q2_i_0(__e); }
            "i_1" => { self._s_Q2_i_1(__e); }
            _ => {}
        }
    }

    fn _state_Q3(&mut self, __e: &MooreMachineFrameEvent) {
        match __e.message.as_str() {
            "$>" => { self._s_Q3_enter(__e); }
            "i_0" => { self._s_Q3_i_0(__e); }
            "i_1" => { self._s_Q3_i_1(__e); }
            _ => {}
        }
    }

    fn _state_Q4(&mut self, __e: &MooreMachineFrameEvent) {
        match __e.message.as_str() {
            "$>" => { self._s_Q4_enter(__e); }
            "i_0" => { self._s_Q4_i_0(__e); }
            "i_1" => { self._s_Q4_i_1(__e); }
            _ => {}
        }
    }

    fn _s_Q0_enter(&mut self, __e: &MooreMachineFrameEvent) {
        self.set_output(0);
    }

    fn _s_Q0_i_1(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q2");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q0_i_0(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q1");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q1_i_1(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q3");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q1_enter(&mut self, __e: &MooreMachineFrameEvent) {
        self.set_output(0);
    }

    fn _s_Q1_i_0(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q1");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q2_i_0(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q4");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q2_i_1(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q2");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q2_enter(&mut self, __e: &MooreMachineFrameEvent) {
        self.set_output(0);
    }

    fn _s_Q3_enter(&mut self, __e: &MooreMachineFrameEvent) {
        self.set_output(1);
    }

    fn _s_Q3_i_1(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q2");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q3_i_0(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q4");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q4_enter(&mut self, __e: &MooreMachineFrameEvent) {
        self.set_output(1);
    }

    fn _s_Q4_i_1(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q3");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn _s_Q4_i_0(&mut self, __e: &MooreMachineFrameEvent) {
        let mut __compartment = MooreMachineCompartment::new("Q1");
        __compartment.parent_compartment = Some(Box::new(self.__compartment.clone()));
        self.__transition(__compartment);
        return;
    }

    fn set_output(&mut self, value: i32) {
                    self.current_output = value;
    }

    pub fn get_output(&mut self) -> i32 {
                    return self.current_output;
    }
}

fn main() {
    println!("TAP version 14");
    println!("1..5");

    let mut m = MooreMachine::new();

    // Initial state Q0 has output 0
    if m.get_output() == 0 {
        println!("ok 1 - moore initial state Q0 has output 0");
    } else {
        println!("not ok 1 - moore initial state Q0 has output 0 # got {}", m.get_output());
    }

    // i_0: Q0 -> Q1 (output 0)
    m.i_0();
    if m.get_output() == 0 {
        println!("ok 2 - moore Q1 has output 0");
    } else {
        println!("not ok 2 - moore Q1 has output 0 # got {}", m.get_output());
    }

    // i_1: Q1 -> Q3 (output 1)
    m.i_1();
    if m.get_output() == 1 {
        println!("ok 3 - moore Q3 has output 1");
    } else {
        println!("not ok 3 - moore Q3 has output 1 # got {}", m.get_output());
    }

    // i_0: Q3 -> Q4 (output 1)
    m.i_0();
    if m.get_output() == 1 {
        println!("ok 4 - moore Q4 has output 1");
    } else {
        println!("not ok 4 - moore Q4 has output 1 # got {}", m.get_output());
    }

    // i_0: Q4 -> Q1 (output 0)
    m.i_0();
    if m.get_output() == 0 {
        println!("ok 5 - moore Q1 has output 0 again");
    } else {
        println!("not ok 5 - moore Q1 has output 0 again # got {}", m.get_output());
    }

    println!("# PASS - Moore machine outputs depend ONLY on state");
}

